--- a/data/org.gnome.Ptyxis.desktop.in.in
+++ b/data/org.gnome.Ptyxis.desktop.in.in
@@ -1,10 +1,10 @@
 [Desktop Entry]
 Name=@APP_NAME@
 Comment=A container-oriented terminal for GNOME
-Exec=ptyxis
+Exec=ptyxis --new-window
 # Translators: Do NOT translate or transliterate this text (this is an icon file name)!
 Icon=@APP_ID@
-DBusActivatable=true
+DBusActivatable=false
 StartupNotify=true
 Terminal=false
 Type=Application
