--- neofetch.orig	2020-08-02 20:37:41.000000000 +0200
+++ neofetch	2021-10-06 21:42:20.251855166 +0200
@@ -6943,25 +6943,27 @@
         ;;
 
         "Fedora"* | "RFRemix"*)
-            set_colors 4 7 1
+            set_colors 12 7
             read -rd '' ascii_data <<'EOF'
-${c1}          /:-------------:\\
-       :-------------------::
-     :-----------${c2}/shhOHbmp${c1}---:\\
-   /-----------${c2}omMMMNNNMMD  ${c1}---:
-  :-----------${c2}sMMMMNMNMP${c1}.    ---:
- :-----------${c2}:MMMdP${c1}-------    ---\\
-,------------${c2}:MMMd${c1}--------    ---:
-:------------${c2}:MMMd${c1}-------    .---:
-:----    ${c2}oNMMMMMMMMMNho${c1}     .----:
-:--     .${c2}+shhhMMMmhhy++${c1}   .------/
-:-    -------${c2}:MMMd${c1}--------------:
-:-   --------${c2}/MMMd${c1}-------------;
-:-    ------${c2}/hMMMy${c1}------------:
-:--${c2} :dMNdhhdNMMNo${c1}------------;
-:---${c2}:sdNMMMMNds:${c1}------------:
-:------${c2}:://:${c1}-------------::
-:---------------------://
+${c1}             .',;::::;,'.
+         .';:cccccccccccc:;,.
+      .;cccccccccccccccccccccc;.
+    .:cccccccccccccccccccccccccc:.
+  .;ccccccccccccc;${c2}.:dddl:.${c1};ccccccc;.
+ .:ccccccccccccc;${c2}OWMKOOXMWd${c1};ccccccc:.
+.:ccccccccccccc;${c2}KMMc${c1};cc;${c2}xMMc${c1}:ccccccc:.
+,cccccccccccccc;${c2}MMM.${c1};cc;${c2};WW:${c1}:cccccccc,
+:cccccccccccccc;${c2}MMM.${c1};cccccccccccccccc:
+:ccccccc;${c2}oxOOOo${c1};${c2}MMM0OOk.${c1};cccccccccccc:
+cccccc:${c2}0MMKxdd:${c1};${c2}MMMkddc.${c1};cccccccccccc;
+ccccc:${c2}XM0'${c1};cccc;${c2}MMM.${c1};cccccccccccccccc'
+ccccc;${c2}MMo${c1};ccccc;${c2}MMW.${c1};ccccccccccccccc;
+ccccc;${c2}0MNc.${c1}ccc${c2}.xMMd${c1}:ccccccccccccccc;
+cccccc;${c2}dNMWXXXWM0:${c1}:cccccccccccccc:,
+cccccccc;${c2}.:odl:.${c1};cccccccccccccc:,.
+:cccccccccccccccccccccccccccc:'.
+.:cccccccccccccccccccccc:;,..
+  '::cccccccccccccc::;,.
 EOF
         ;;
 
