diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0df659a..6251793 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -27,9 +27,6 @@ endif()
 
 find_package(Qt6 COMPONENTS Widgets Core)
 target_link_libraries(${PROJECT_NAME} PRIVATE Qt6::Core Qt6::Widgets)
-if(OS_LINUX OR OS_FREEBSD OR OS_OPENBSD)
-	target_link_libraries(${PROJECT_NAME} PRIVATE Qt::GuiPrivate)
-endif()
 
 target_compile_options(
 ${PROJECT_NAME} PRIVATE $<$<C_COMPILER_ID:Clang,AppleClang>:-Wno-quoted-include-in-framework-header
