diff -up chromium-98.0.4758.80/remoting/base/util.cc.remoting-cstring chromium-98.0.4758.80/remoting/base/util.cc
--- chromium-98.0.4758.80/remoting/base/util.cc.remoting-cstring	2022-02-01 00:41:54.000000000 +0000
+++ chromium-98.0.4758.80/remoting/base/util.cc	2022-02-02 15:55:01.352340901 +0000
@@ -4,6 +4,7 @@
 
 #include "remoting/base/util.h"
 
+#include <cstring>
 #include <math.h>
 
 #include "base/logging.h"
diff -up chromium-98.0.4758.80/remoting/client/display/gl_helpers.cc.remoting-cstring chromium-98.0.4758.80/remoting/client/display/gl_helpers.cc
--- chromium-98.0.4758.80/remoting/client/display/gl_helpers.cc.remoting-cstring	2022-02-01 00:41:54.000000000 +0000
+++ chromium-98.0.4758.80/remoting/client/display/gl_helpers.cc	2022-02-02 15:55:01.353340854 +0000
@@ -5,6 +5,7 @@
 #include "remoting/client/display/gl_helpers.h"
 
 #include "base/logging.h"
+#include <cstring>
 
 namespace remoting {
 
diff -up chromium-98.0.4758.80/remoting/host/fake_mouse_cursor_monitor.cc.remoting-cstring chromium-98.0.4758.80/remoting/host/fake_mouse_cursor_monitor.cc
--- chromium-98.0.4758.80/remoting/host/fake_mouse_cursor_monitor.cc.remoting-cstring	2022-02-01 00:41:54.000000000 +0000
+++ chromium-98.0.4758.80/remoting/host/fake_mouse_cursor_monitor.cc	2022-02-02 15:55:01.353340854 +0000
@@ -4,6 +4,7 @@
 
 #include "remoting/host/fake_mouse_cursor_monitor.h"
 
+#include <cstring>
 #include <memory>
 
 #include "base/check.h"
diff -up chromium-98.0.4758.80/remoting/host/host_config_unittest.cc.remoting-cstring chromium-98.0.4758.80/remoting/host/host_config_unittest.cc
--- chromium-98.0.4758.80/remoting/host/host_config_unittest.cc.remoting-cstring	2022-02-02 15:55:01.353340854 +0000
+++ chromium-98.0.4758.80/remoting/host/host_config_unittest.cc	2022-02-02 15:57:20.295794172 +0000
@@ -10,6 +10,7 @@
 #include "base/values.h"
 #include "testing/gtest/include/gtest/gtest.h"
 #include "third_party/abseil-cpp/absl/types/optional.h"
+#include <cstring>
 
 namespace remoting {
 
diff -up chromium-98.0.4758.80/remoting/host/host_secret.cc.remoting-cstring chromium-98.0.4758.80/remoting/host/host_secret.cc
--- chromium-98.0.4758.80/remoting/host/host_secret.cc.remoting-cstring	2022-02-01 00:41:54.000000000 +0000
+++ chromium-98.0.4758.80/remoting/host/host_secret.cc	2022-02-02 15:55:01.353340854 +0000
@@ -6,7 +6,7 @@
 
 #include <stdint.h>
 
-#include <string>
+#include <cstring>
 
 #include "base/rand_util.h"
 #include "base/strings/string_number_conversions.h"
