diff --git a/src/kcm/options.kcfg b/src/kcm/options.kcfg
index 8b64b57..b079008 100644
--- a/src/kcm/options.kcfg
+++ b/src/kcm/options.kcfg
@@ -78,12 +78,12 @@
             <default>black</default>
         </entry>
         <entry name="ActiveSecondOutlineAlpha" type="Int">
-            <default>255</default>
+            <default>70</default>
             <min>0</min>
             <max>255</max>
         </entry>
         <entry name="InactiveSecondOutlineAlpha" type="Int">
-            <default>255</default>
+            <default>0</default>
             <min>0</min>
             <max>255</max>
         </entry>
