diff --git a/src/kcm/options.kcfg b/src/kcm/options.kcfg
index 813fcc4..d1c838b 100644
--- a/src/kcm/options.kcfg
+++ b/src/kcm/options.kcfg
@@ -43,10 +43,10 @@
             <max>20</max>
         </entry>
         <entry name="OutlineColor" type="Color">
-            <default>black</default>
+            <default>62,62,62</default>
         </entry>
         <entry name="InactiveOutlineColor" type="Color">
-            <default>black</default>
+            <default>62,62,62</default>
         </entry>
         <entry name="ActiveOutlineAlpha" type="Int">
             <default>255</default>
@@ -80,12 +80,12 @@
 
     <group name="Shadow">
         <entry name="ActiveShadowAlpha" type="Int">
-            <default>255</default>
+            <default>70</default>
             <min>0</min>
             <max>255</max>
         </entry>
         <entry name="InactiveShadowAlpha" type="Int">
-            <default>255</default>
+            <default>0</default>
             <min>0</min>
             <max>255</max>
         </entry>
@@ -96,16 +96,16 @@
             <default>11</default>
         </entry>
         <entry name="ActiveShadowUsePalette" type="Bool">
-            <default>true</default>
+            <default>false</default>
         </entry>
         <entry name="InactiveShadowUsePalette" type="Bool">
-            <default>true</default>
+            <default>false</default>
         </entry>
         <entry name="ActiveShadowUseCustom" type="Bool">
-            <default>false</default>
+            <default>true</default>
         </entry>
         <entry name="InactiveShadowUseCustom" type="Bool">
-            <default>false</default>
+            <default>true</default>
         </entry>
     </group>
 
