diff --git a/src/kcm/options.kcfg b/src/kcm/options.kcfg
index cf4c021..fbbf4f3 100644
--- a/src/kcm/options.kcfg
+++ b/src/kcm/options.kcfg
@@ -132,12 +132,12 @@
             <default>white</default>
         </entry>
         <entry name="ActiveSecondOutlineAlpha" type="Int">
-            <default>85</default>
+            <default>0</default>
             <min>0</min>
             <max>255</max>
         </entry>
         <entry name="InactiveSecondOutlineAlpha" type="Int">
-            <default>85</default>
+            <default>0</default>
             <min>0</min>
             <max>255</max>
         </entry>
