diff --git a/src/kcm/options.kcfg b/src/kcm/options.kcfg
index 6354fe3..193f986 100644
--- a/src/kcm/options.kcfg
+++ b/src/kcm/options.kcfg
@@ -41,10 +41,10 @@
             <max>20</max>
         </entry>
         <entry name="OutlineColor" type="Color">
-            <default>black</default>
+            <default>62,62,62</default>
         </entry>
         <entry name="InactiveOutlineColor" type="Color">
-            <default>black</default>
+            <default>62,62,62</default>
         </entry>
         <entry name="ActiveOutlineAlpha" type="Int">
             <default>255</default>
@@ -77,12 +77,12 @@
 
 
         <entry name="ActiveShadowAlpha" type="Int">
-            <default>255</default>
+            <default>70</default>
             <min>0</min>
             <max>255</max>
         </entry>
         <entry name="InactiveShadowAlpha" type="Int">
-            <default>255</default>
+            <default>0</default>
             <min>0</min>
             <max>255</max>
         </entry>
@@ -93,16 +93,16 @@
             <default>11</default>
         </entry>
         <entry name="ActiveShadowUsePalette" type="Bool">
-            <default>true</default>
+            <default>false</default>
         </entry>
         <entry name="InactiveShadowUsePalette" type="Bool">
-            <default>true</default>
+            <default>false</default>
         </entry>
         <entry name="ActiveShadowUseCustom" type="Bool">
-            <default>false</default>
+            <default>true</default>
         </entry>
         <entry name="InactiveShadowUseCustom" type="Bool">
-            <default>false</default>
+            <default>true</default>
         </entry>
 
 
@@ -131,4 +131,4 @@
             <default>true</default>
         </entry>
     </group>
-</kcfg>
\ No newline at end of file
+</kcfg>
