diff --git a/components/calendar/kconf_update/migrate-calendar-to-plugin-id.py b/components/calendar/kconf_update/migrate-calendar-to-plugin-id.py
index 1e3b710887..bc4d150fad 100644
--- a/components/calendar/kconf_update/migrate-calendar-to-plugin-id.py
+++ b/components/calendar/kconf_update/migrate-calendar-to-plugin-id.py
@@ -12,7 +12,7 @@ import subprocess
 
 proc = subprocess.Popen(
     [
-        "qtpaths",
+        "qtpaths-qt6",
         "--locate-file",
         "ConfigLocation",
         "plasma-org.kde.plasma.desktop-appletsrc",
