From a5503d5a6ac3e65488e936b8f67062025f231c72 Mon Sep 17 00:00:00 2001
From: Samuel Pitoiset <samuel.pitoiset@gmail.com>
Date: Tue, 27 Feb 2024 20:33:10 +0100
Subject: [PATCH] hack2

Signed-off-by: Samuel Pitoiset <samuel.pitoiset@gmail.com>
---
 src/amd/vulkan/radv_device.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/amd/vulkan/radv_device.c b/src/amd/vulkan/radv_device.c
index 402341703e693..dcedc1bc59292 100644
--- a/src/amd/vulkan/radv_device.c
+++ b/src/amd/vulkan/radv_device.c
@@ -1066,6 +1066,21 @@ radv_CreateDevice(VkPhysicalDevice physicalDevice, const VkDeviceCreateInfo *pCr
    if (result != VK_SUCCESS)
       goto fail_cache;
 
+   {
+      struct radeon_winsys *ws = device->ws;
+
+      if (device->physical_device->rad_info.has_stable_pstate) {
+         for (unsigned i = 0; i < RADV_NUM_HW_CTX; i++) {
+            if (!device->hw_ctx[i])
+               continue;
+
+            if (ws->ctx_set_pstate(device->hw_ctx[i], RADEON_CTX_PSTATE_STANDARD) < 0) {
+               fprintf(stderr, "radv: failed to set pstate to profile_standard\n");
+            }
+         }
+      }
+   }
+
    *pDevice = radv_device_to_handle(device);
    return VK_SUCCESS;
 
-- 
GitLab

