diff --git a/bin_steam.sh b/bin_steam.sh
index fb81d47..0d04291 100755
--- a/bin_steam.sh
+++ b/bin_steam.sh
@@ -26,6 +26,33 @@ export STEAMSCRIPT_VERSION=1.0.0.79
 # Set up domain for script localization
 export TEXTDOMAIN=steam
 
+# Initialize an empty array to hold the arguments
+args=()
+
+# Loop through all arguments
+while [ "$#" -gt 0 ]; do
+    case "$1" in
+        -steamdeck)
+            # If -steamdeck is found, just add it to the args array
+            args+=("$1")
+            ;;
+        *)
+            # For all other arguments, add them to the args array
+            args+=("$1")
+            ;;
+    esac
+    # Shift the arguments to process the next one
+    shift
+done
+
+# Check if -steamdeck is in the args array
+if [[ ! " ${args[@]} " =~ " -steamdeck " ]]; then
+    # If -steamdeck is not found, append it
+    args+=("-steamdeck")
+fi
+set -- "${args[@]}"
+
+
 function show_message()
 {
 	style=$1
