diff --git a/bin_steam.sh b/bin_steam.sh
index fb81d47..cf38071 100755
--- a/bin_steam.sh
+++ b/bin_steam.sh
@@ -133,18 +133,22 @@ function install_bootstrap()
 	if [ -d "$DESKTOP_DIR" ] && [ "$bootstrapdir" = "/usr/lib/$STEAMPACKAGE" ]; then
 		# There might be a symlink in place already, in such case we do nothing
 		if [ ! -L "$DESKTOP_DIR/$STEAMPACKAGE.desktop" ]; then
-			cp "$bootstrapdir/$STEAMPACKAGE.desktop" "$DESKTOP_DIR"
+			cp "/usr/share/applications/$STEAMPACKAGE.desktop" "$DESKTOP_DIR"
+			cp "/usr/share/applications/RemoteHost.desktop" "$DESKTOP_DIR"
 			# Older .desktop implementations used the execute bits as
 			# a marker for a .desktop being safe to treat as a shortcut
 			chmod a+x "$DESKTOP_DIR/$STEAMPACKAGE.desktop"
+			chmod a+x "$DESKTOP_DIR/RemoteHost.desktop"
 			if command -v gio >/dev/null; then
 				# Making it executable is not enough in recent
 				# (Ubuntu 20.04) versions of
 				# https://gitlab.gnome.org/World/ShellExtensions/desktop-icons
 				gio set --type=string "$DESKTOP_DIR/$STEAMPACKAGE.desktop" metadata::trusted true || :
+				gio set --type=string "$DESKTOP_DIR/RemoteHost.desktop" metadata::trusted true || :
 				# Generate an inotify event so the desktop
 				# implementation reloads it
 				touch "$DESKTOP_DIR/$STEAMPACKAGE.desktop"
+				touch "$DESKTOP_DIR/RemoteHost.desktop"
 			fi
 		fi
 	fi
