diff --git a/bin_steam.sh b/bin_steam.sh
index a88deda..d890785 100755
--- a/bin_steam.sh
+++ b/bin_steam.sh
@@ -44,6 +44,13 @@ if [ "$steamdeck_present" == false ]; then
     set -- "$@" "-steamdeck"
 fi

+# Use wayland compatible libextest
+if [[ "$XDG_SESSION_TYPE" == "wayland" ]]; then
+  if lsusb | grep -q Valve; then
+    export LD_PRELOAD=/usr/lib/steam/libextest.so
+  fi
+fi
+
 function show_message()
 {
 	style=$1
