diff --git a/lookandfeel/sddm-theme/Login.qml b/lookandfeel/sddm-theme/Login.qml
index 24a5b42..ce31749 100644
--- a/lookandfeel/sddm-theme/Login.qml
+++ b/lookandfeel/sddm-theme/Login.qml
@@ -91,7 +91,9 @@ SessionManagementScreen {
 
             placeholderText: i18nd("plasma_lookandfeel_org.kde.lookandfeel", "Password")
             focus: !showUsernamePrompt || lastUserName
-
+            echoMode: TextInput.Password
+            revealPasswordButtonShown: true
+        
             // Disable reveal password action because SDDM does not have the breeze icon set loaded
             rightActions: []
 
