diff -bur plasma-workspace-5.27.3.old/lookandfeel/sddm-theme/Main.qml plasma-workspace-5.27.3/lookandfeel/sddm-theme/Main.qml
--- plasma-workspace-5.27.3.old/lookandfeel/sddm-theme/Main.qml	2023-03-14 15:12:21.000000000 +0100
+++ plasma-workspace-5.27.3/lookandfeel/sddm-theme/Main.qml	2023-03-27 20:20:10.759761866 +0200
@@ -555,7 +555,7 @@
                     userListComponent.mainPasswordBox.forceActiveFocus();
                     inputPanel.showHide()
                 }
-                visible: inputPanel.status === Loader.Ready
+                visible: false
             }
 
             KeyboardButton {
