diff --git a/packaging/linux/sunshine.service.in b/packaging/linux/sunshine.service.in
index 1ba3a13..9ea0926 100644
--- a/packaging/linux/sunshine.service.in
+++ b/packaging/linux/sunshine.service.in
@@ -2,6 +2,7 @@
 Description=@PROJECT_DESCRIPTION@
 StartLimitIntervalSec=500
 StartLimitBurst=5
+PartOf=xdg-desktop-portal.service
 
 [Service]
 # Avoid starting Sunshine before the desktop is fully initialized.
