diff --git a/data/org.gnome.mutter.gschema.xml.in b/data/org.gnome.mutter.gschema.xml.in
index c014b749f..b2b48626f 100644
--- a/data/org.gnome.mutter.gschema.xml.in
+++ b/data/org.gnome.mutter.gschema.xml.in
@@ -102,7 +102,7 @@
     </key>
 
     <key name="center-new-windows" type="b">
-      <default>false</default>
+      <default>true</default>
       <summary>Place new windows in the center</summary>
       <description>
         When true, the new windows will always be put in the center of the
