diff --git a/data/org.gnome.mutter.gschema.xml.in b/data/org.gnome.mutter.gschema.xml.in
index c014b749f..2925f9e10 100644
--- a/data/org.gnome.mutter.gschema.xml.in
+++ b/data/org.gnome.mutter.gschema.xml.in
@@ -154,7 +154,7 @@
     </key>
 
     <key name="check-alive-timeout" type="u">
-      <default>5000</default>
+      <default>30000</default>
       <summary>Timeout for check-alive ping</summary>
       <description>
         Number of milliseconds a client has to respond to a ping request in
