diff --git a/gnome-initial-setup/pages/privacy/gis-privacy-page.c b/gnome-initial-setup/pages/privacy/gis-privacy-page.c
index 1e3934e..54e38f0 100644
--- a/gnome-initial-setup/pages/privacy/gis-privacy-page.c
+++ b/gnome-initial-setup/pages/privacy/gis-privacy-page.c
@@ -119,7 +119,7 @@ gis_privacy_page_constructed (GObject *object)
   page->privacy_settings = g_settings_new ("org.gnome.desktop.privacy");
 
   gtk_switch_set_active (GTK_SWITCH (page->location_switch), TRUE);
-  gtk_switch_set_active (GTK_SWITCH (page->reporting_switch), TRUE);
+  gtk_switch_set_active (GTK_SWITCH (page->reporting_switch), FALSE);
 
   gtk_label_set_label (GTK_LABEL (page->location_privacy_label),
                        _("Allows apps to determine your geographical location."));
