diff --git a/gnome-initial-setup/pages/privacy/gis-privacy-page.c b/gnome-initial-setup/pages/privacy/gis-privacy-page.c
index b910651..10663e8 100644
--- a/gnome-initial-setup/pages/privacy/gis-privacy-page.c
+++ b/gnome-initial-setup/pages/privacy/gis-privacy-page.c
@@ -121,7 +121,7 @@ gis_privacy_page_constructed (GObject *object)
   priv->privacy_settings = g_settings_new ("org.gnome.desktop.privacy");

   gtk_switch_set_active (GTK_SWITCH (priv->location_switch), TRUE);
-  gtk_switch_set_active (GTK_SWITCH (priv->reporting_switch), TRUE);
+  gtk_switch_set_active (GTK_SWITCH (priv->reporting_switch), FALSE);

   if (update_os_data (page))
     {
--
2.41.0

