diff --git a/src/hhd/device/rog_ally/base.py b/src/hhd/device/rog_ally/base.py
index 001086d..30d062d 100644
--- a/src/hhd/device/rog_ally/base.py
+++ b/src/hhd/device/rog_ally/base.py
@@ -87,19 +87,19 @@ ALLY_X_BUTTON_MAP: dict[int, GamepadButton] = to_map(
         # Gamepad
         "a": [EC("BTN_SOUTH")],
         "b": [EC("BTN_EAST")],
-        "x": [EC("BTN_C")],
-        "y": [EC("BTN_NORTH")],
+        "x": [EC("BTN_NORTH")],
+        "y": [EC("BTN_WEST")],
         # Sticks
-        "ls": [EC("BTN_TL2")],
-        "rs": [EC("BTN_TR2")],
+        "ls": [EC("BTN_THUMBL")],
+        "rs": [EC("BTN_THUMBR")],
         # Bumpers
-        "lb": [EC("BTN_WEST")],
-        "rb": [EC("BTN_Z")],
+        "lb": [EC("BTN_TL")],
+        "rb": [EC("BTN_TR")],
         # Select
-        "start": [EC("BTN_TR")],
-        "select": [EC("BTN_TL")],
+        "start": [EC("BTN_START")],
+        "select": [EC("BTN_SELECT")],
         # Misc
-        # "mode": [EC("BTN_MODE")],
+        "mode": [EC("BTN_MODE")],
     }
 )
 
