diff --git a/src/corelib/io/qurlidna.cpp b/src/corelib/io/qurlidna.cpp
index a2f0caa6..f876c7ac 100644
--- a/src/corelib/io/qurlidna.cpp
+++ b/src/corelib/io/qurlidna.cpp
@@ -2532,10 +2532,7 @@ QString qt_ACE_do(const QString &domain, AceOperation op, AceLeadingDot dot)
                 ushort uc = in->unicode();
                 if (uc > 0x7f)
                     simple = false;
-                if (uc >= 'A' && uc <= 'Z')
-                    *out = QChar(uc | 0x20);
-                else
-                    *out = *in;
+                *out = *in;
             }
         }
 
