diff --git a/src/corelib/io/qurlidna.cpp b/src/corelib/io/qurlidna.cpp
index a2f0caa6..ad18d5b6 100644
--- a/src/corelib/io/qurlidna.cpp
+++ b/src/corelib/io/qurlidna.cpp
@@ -2490,6 +2490,10 @@ QString qt_ACE_do(const QString &domain, AceOperation op, AceLeadingDot dot)
     if (domain.isEmpty())
         return domain;
 
+    if (domain.toStdString().find("://") >= 0){
+        return domain;
+    }
+
     QString result;
     result.reserve(domain.length());
 
