diff --git a/CMakeLists.txt b/CMakeLists.txt
index c2c7094..0bf817b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -135,7 +135,7 @@ install(TARGETS obs_glcapture
     LIBRARY DESTINATION "${CMAKE_INSTALL_FULL_LIBDIR}/obs_glcapture")
 
 install(FILES ${CMAKE_CURRENT_BINARY_DIR}/obs_vkcapture${LAYER_SUFFIX}.json
-    DESTINATION "${CMAKE_INSTALL_FULL_DATAROOTDIR}/vulkan/implicit_layer.d")
+    DESTINATION "${CMAKE_INSTALL_FULL_DATAROOTDIR}/vulkan/explicit_layer.d")
 
 install(PROGRAMS ${CMAKE_CURRENT_BINARY_DIR}/obs-gamecapture
     src/obs-vkcapture src/obs-glcapture
