From a85e11871c748eae11d056d93df644e2dcb1e80c Mon Sep 17 00:00:00 2001
From: BenzhengZhang <140143892+BenzhengZhang@users.noreply.github.com>
Date: Fri, 14 Feb 2025 11:19:27 +0800
Subject: [PATCH] Update shared lib version for meson (#3860)

---
 meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 6e8b5854..573f6c60 100644
--- a/meson.build
+++ b/meson.build
@@ -4,7 +4,7 @@ project('openh264', ['c', 'cpp'],
   default_options : [ 'warning_level=1',
                       'buildtype=debugoptimized' ])
 
-major_version = '7'
+major_version = '8'
 
 cpp = meson.get_compiler('cpp')
 
-- 
2.48.1

