diff --git a/data/org.gnome.gedit.gschema.xml.in b/data/org.gnome.gedit.gschema.xml.in
index 59325f4..263999a 100644
--- a/data/org.gnome.gedit.gschema.xml.in
+++ b/data/org.gnome.gedit.gschema.xml.in
@@ -43,7 +43,7 @@
       <description>A custom font that will be used for the editing area. This will only take effect if the “Use Default Font” option is turned off.</description>
     </key>
     <key name="scheme" type="s">
-      <default>'tango'</default>
+      <default>'oblivion'</default>
       <summary>Style Scheme</summary>
       <description>The ID of a GtkSourceView Style Scheme used to color the text.</description>
     </key>
