diff --git a/data/org.gnome.gedit.gschema.xml.in b/data/org.gnome.gedit.gschema.xml.in
index 8ea1f0b..f03fd93 100644
--- a/data/org.gnome.gedit.gschema.xml.in
+++ b/data/org.gnome.gedit.gschema.xml.in
@@ -49,7 +49,7 @@
       <description>The ID of a libgedit-gtksourceview style scheme used to color the main text area when the application has a light theme applied.</description>
     </key>
     <key name="style-scheme-for-dark-theme-variant" type="s">
-      <default>'solarized-dark'</default>
+      <default>'oblivion'</default>
       <summary>Style Scheme for Dark Theme Variant</summary>
       <description>The ID of a libgedit-gtksourceview style scheme used to color the main text area when the application has a dark theme applied.</description>
     </key>
