From 30649e5a643e7d780a17196fed931de3443efa8d Mon Sep 17 00:00:00 2001
From: GloriousEggroll <gloriouseggroll@gmail.com>
Date: Tue, 4 Oct 2022 13:02:15 -0600
Subject: [PATCH 2/3] Don't create a separate /home partition for any default
 schemes

---
 data/anaconda.conf | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/data/anaconda.conf b/data/anaconda.conf
index 6624da5..0dc87d4 100644
--- a/data/anaconda.conf
+++ b/data/anaconda.conf
@@ -200,8 +200,7 @@ file_system_type =
 #   btrfs          The mount point will be created only for the Btrfs scheme
 #
 default_partitioning =
-    /     (min 1 GiB, max 70 GiB)
-    /home (min 500 MiB, free 50 GiB)
+    /     (min 1 GiB)
 
 # Default partitioning scheme.
 # Valid values:
-- 
2.38.1

