diff --git a/CMakeLists.txt b/CMakeLists.txt
index 10882e8..e2c4f44 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,4 +1,13 @@
-cmake_minimum_required(VERSION 3.16...3.21)
+# --- Detect if the plugin is build out of tree or not ---
+if(CMAKE_PROJECT_NAME STREQUAL "obs-studio")
+  set(BUILD_OUT_OF_TREE OFF)
+  if(OBS_CMAKE_VERSION VERSION_GREATER_EQUAL 3.0.0)
+  	legacy_check()
+  endif()
+else()
+  set(BUILD_OUT_OF_TREE ON)
+  cmake_minimum_required(VERSION 3.16...3.21)
+endif()
 
 # Change obs-plugintemplate to your plugin's name in a machine-readable format (e.g.:
 # obs-myawesomeplugin) and set
@@ -29,8 +38,14 @@ target_sources(${CMAKE_PROJECT_NAME} PRIVATE src/shuffler.h)
 target_sources(${CMAKE_PROJECT_NAME} PRIVATE src/shuffler.c)
 
 # Import libobs as main plugin dependency
-find_package(libobs REQUIRED)
-include(cmake/ObsPluginHelpers.cmake)
+if(BUILD_OUT_OF_TREE)
+	find_package(libobs REQUIRED)
+	find_package(obs-frontend-api REQUIRED)
+	include(cmake/ObsPluginHelpers.cmake)
+	set(OBS_FRONTEND_API_NAME "obs-frontend-api")
+else()
+	set(OBS_FRONTEND_API_NAME "frontend-api")
+endif()
 
 # Uncomment these lines if you want to use the OBS Frontend API in your plugin
 #[[
