diff --git a/lookandfeel/sddm-theme/Main.qml b/lookandfeel/sddm-theme/Main.qml
index f93d3261cf..cb823f3a2b 100644
--- a/lookandfeel/sddm-theme/Main.qml
+++ b/lookandfeel/sddm-theme/Main.qml
@@ -572,7 +572,7 @@ Item {
                     userListComponent.mainPasswordBox.forceActiveFocus();
                     inputPanel.showHide()
                 }
-                visible: inputPanel.status === Loader.Ready
+                visible: false
 
                 Layout.fillHeight: true
                 containmentMask: Item {
