--- a/src/main.conf	2024-05-16 13:55:28.987362698 +0100
+++ b/src/main.conf	2024-05-16 13:56:11.115717819 +0100
@@ -137,7 +137,7 @@
 # a6695ace-ee7f-4fb9-881a-5fac66c629af (BlueZ Experimental Offload Codecs)
 # 6fbaf188-05e0-496a-9885-d6ddfdb4e03e (BlueZ Experimental ISO socket)
 # Defaults to false.
-#KernelExperimental = false
+KernelExperimental = 15c0a148-c273-11ea-b3de-0242ac130004
 
 # The duration to avoid retrying to resolve a peer's name, if the previous
 # try failed.
