From 0291d498c0e1c7452dd4fa15ce3088a5cb92e2d6 Mon Sep 17 00:00:00 2001
From: GloriousEggroll <gloriouseggroll@gmail.com>
Date: Mon, 27 Mar 2023 15:54:40 -0600
Subject: [PATCH] disable zchunk for Nobara snapshots

---
 etc/dnf/dnf.conf | 1 +
 1 file changed, 1 insertion(+)

diff --git a/etc/dnf/dnf.conf b/etc/dnf/dnf.conf
index 1058a27..d503c4f 100644
--- a/etc/dnf/dnf.conf
+++ b/etc/dnf/dnf.conf
@@ -7,4 +7,5 @@ clean_requirements_on_remove=True
 best=False
 skip_if_unavailable=True
 max_parallel_downloads=6
+zchunk=False
 
-- 
2.39.2

