From b70be537219c9ddbac012c26c42f628225eefa59 Mon Sep 17 00:00:00 2001
From: Thomas Crider <gloriouseggroll@gmail.com>
Date: Wed, 18 May 2022 14:13:48 -0600
Subject: [PATCH] increase parallel downloads
 downloads to 6

---
 etc/dnf/dnf.conf | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/etc/dnf/dnf.conf b/etc/dnf/dnf.conf
index 460dd25..9fa4a23 100644
--- a/etc/dnf/dnf.conf
+++ b/etc/dnf/dnf.conf
@@ -6,3 +6,5 @@ installonly_limit=3
 clean_requirements_on_remove=True
 best=False
 skip_if_unavailable=True
+max_parallel_downloads=6
+
-- 
2.36.1

