From c437d33f9b034b92de8ca0e211711ffa3df88d49 Mon Sep 17 00:00:00 2001
From: Thomas Crider <tcrider@localhost.localdomain>
Date: Wed, 16 Nov 2022 12:37:47 -0700
Subject: [PATCH 3/6] Revert 'elf: Generalize name-based DSO recognition in
 ldconfig'

---
 elf/Makefile   | 1 -
 elf/ldconfig.c | 5 +++--
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/elf/Makefile b/elf/Makefile
index 6cdd03e4..f3898b52 100644
--- a/elf/Makefile
+++ b/elf/Makefile
@@ -399,7 +399,6 @@ tests += \
   tst-big-note \
   tst-debug1 \
   tst-deep1 \
-  tst-dl-is_dso \
   tst-dlmodcount \
   tst-dlmopen1 \
   tst-dlmopen3 \
diff --git a/elf/ldconfig.c b/elf/ldconfig.c
index 166dccb5..8d7ddf7b 100644
--- a/elf/ldconfig.c
+++ b/elf/ldconfig.c
@@ -42,7 +42,6 @@
 #include <ldconfig.h>
 #include <dl-cache.h>
 #include <dl-hwcaps.h>
-#include <dl-is_dso.h>
 
 #include <dl-procinfo.h>
 
@@ -716,7 +715,9 @@ search_dir (const struct dir_entry *entry)
 	continue;
       /* Does this file look like a shared library?  The dynamic linker
 	 is also considered as shared library.  */
-      if (!_dl_is_dso (direntry->d_name)
+      if (((strncmp (direntry->d_name, "lib", 3) != 0
+	    && strncmp (direntry->d_name, "ld-", 3) != 0)
+	   || strstr (direntry->d_name, ".so") == NULL)
 	  && (direntry->d_type == DT_REG || entry->hwcaps == NULL))
 	continue;
 
-- 
2.39.2

