From 52e498a1d4ab6f20048b613e0a27037989a3bec8 Mon Sep 17 00:00:00 2001
From: Thomas Crider <tcrider@localhost.localdomain>
Date: Wed, 16 Nov 2022 12:39:52 -0700
Subject: [PATCH 5/6] Revert 'nptl_db: Install libthread_db under a regular
 implementation name'

---
 nptl_db/Makefile | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/nptl_db/Makefile b/nptl_db/Makefile
index 2360e53c..c5b48c0b 100644
--- a/nptl_db/Makefile
+++ b/nptl_db/Makefile
@@ -21,6 +21,8 @@ subdir          := nptl_db
 
 include ../Makeconfig
 
+nptl_db-version = 1.0
+
 extra-libs = libthread_db
 extra-libs-others := $(extra-libs)
 
-- 
2.38.1

