diff --git a/src/input/target/steam_deck.rs b/src/input/target/steam_deck.rs
index 6e1674e..d198981 100644
--- a/src/input/target/steam_deck.rs
+++ b/src/input/target/steam_deck.rs
@@ -62,27 +62,17 @@ pub enum ProductId {
     //LenovoLegionGo2 = 0x13__,
 }
 
+//just a hack for now so we get steam guide button icon for all devices
 impl ProductId {
     pub fn to_u16(&self) -> u16 {
         match self {
-            ProductId::SteamDeck => ProductId::SteamDeck as u16,
-            ProductId::Generic => ProductId::Generic as u16,
-            ProductId::ZotacZone => ProductId::ZotacZone as u16,
-            ProductId::AsusRogAlly => ProductId::AsusRogAlly as u16,
-            ProductId::LenovoLegionGo => ProductId::LenovoLegionGo as u16,
-            ProductId::LenovoLegionGoS => ProductId::LenovoLegionGoS as u16,
+            ProductId::LenovoLegionGoS => 0x12ff,
+            _ => 0x12f0,
         }
     }
 
     pub fn to_u32(&self) -> u32 {
-        match self {
-            ProductId::SteamDeck => ProductId::SteamDeck as u32,
-            ProductId::Generic => ProductId::Generic as u32,
-            ProductId::ZotacZone => ProductId::ZotacZone as u32,
-            ProductId::AsusRogAlly => ProductId::AsusRogAlly as u32,
-            ProductId::LenovoLegionGo => ProductId::LenovoLegionGo as u32,
-            ProductId::LenovoLegionGoS => ProductId::LenovoLegionGoS as u32,
-        }
+        self.to_u16() as u32
     }
 }
 
