diff --git a/Makefile b/Makefile
index 86917b8..65cc1ca 100644
--- a/Makefile
+++ b/Makefile
@@ -8,14 +8,15 @@ build:
 
 install:
 	sed "s:^PREFIX=\"/usr\":PREFIX=\"$(PREFIX)\":" -i steamtinkerlaunch
-	install -Dm755 steamtinkerlaunch -t "$(PREFIX)/bin"
+	install -Dm755 steamtinkerlaunch -t "$(BUILDROOT)$(PREFIX)/bin"
 
-	install -d "$(PREFIX)/share/steamtinkerlaunch"
-	cp -r collections eval guicfgs lang misc "$(PREFIX)/share/steamtinkerlaunch"
+	install -d "$(BUILDROOT)$(PREFIX)/share/steamtinkerlaunch"
+	cp -r collections eval guicfgs lang misc "$(BUILDROOT)$(PREFIX)/share/steamtinkerlaunch"
 
-	install -Dm644 README.md -t "$(PREFIX)/share/doc/steamtinkerlaunch"
-	install -Dm644 "misc/steamtinkerlaunch.desktop" -t "$(PREFIX)/share/applications"
-	install -Dm644 "misc/steamtinkerlaunch.svg" -t "$(PREFIX)/share/icons/hicolor/scalable/apps"
+	install -Dm644 README.md -t "$(BUILDROOT)$(PREFIX)/share/doc/steamtinkerlaunch"
+	install -Dm644 "misc/steamtinkerlaunch.desktop" -t "$(BUILDROOT)$(PREFIX)/share/applications"
+	install -Dm644 "misc/steamtinkerlaunch.svg" -t "$(BUILDROOT)$(PREFIX)/share/icons/hicolor/scalable/apps"
+	install -Dm644 "misc/steamtinkerlaunch.svg" -t "$(BUILDROOT)$(PREFIX)/share/steamtinkerlaunch/misc"
 
 uninstall:
 	rm -rf "${PREFIX}/share/icons/hicolor/scalable/apps/misc/steamtinkerlaunch.svg"
