diff --git a/src/gallium/drivers/radeonsi/radeon_vcn_enc.c b/src/gallium/drivers/radeonsi/radeon_vcn_enc.c
index 1a19f340d1f..fe122f1f0e5 100644
--- a/src/gallium/drivers/radeonsi/radeon_vcn_enc.c
+++ b/src/gallium/drivers/radeonsi/radeon_vcn_enc.c
@@ -1024,6 +1024,8 @@ static void radeon_vcn_enc_av1_get_param(struct radeon_encoder *enc,
       enc_pic->av1_enc_params.lsm_reference_frame_index[1] = 0xffffffff;
       enc_pic->av1.compound = false;
 
+      pic->ref_list0[1] = PIPE_H2645_LIST_REF_INVALID_ENTRY;
+
       if (pic->ref_list1[0] != PIPE_H2645_LIST_REF_INVALID_ENTRY) {
          enc_pic->av1.compound = true; /* BIDIR_COMP */
          enc_pic->av1_enc_params.lsm_reference_frame_index[1] = pic->ref_list1[0];

