diff --git a/profile b/profile
index de8e9ab..e3f75bf 100644
--- a/profile
+++ b/profile
@@ -83,3 +83,49 @@ then
    . /etc/bashrc
 fi
 
+# workaround for input issue https://github.com/ibus/ibus/issues/2618
+export IBUS_ENABLE_SYNC_MODE=2
+
+# nvidia rendering fixup
+if [ -f /bin/lspci ]; then
+nvgpu=$(lspci | grep -iE 'VGA|3D' | grep -i nvidia | cut -d ":" -f 3)
+
+if [[ ! -z $nvgpu ]]; then
+    export LIBVA_DRIVER_NAME=nvidia
+    export MOZ_DISABLE_RDD_SANDBOX=1
+    export EGL_PLATFORM=$XDG_SESSION_TYPE
+fi
+fi
+
+# fix gnome missing 'New file' option
+if [ ! -f $HOME/Templates/"Text file" ]
+then
+    mkdir -p $HOME/Templates
+    touch $HOME/Templates/"Text file"
+fi
+
+# enable vulkan capture layer globally by default
+export OBS_VKCAPTURE=1
+export OBS_VKCAPTURE_QUIET=1
+# this is a hack to disable Discord from trying to update itself. We are only able to receive discord updates from our package repo.
+if [ ! -f $HOME/.config/discord/settings.json ];then
+    mkdir -p $HOME/.config/discord/
+    echo '{' > $HOME/.config/discord/settings.json
+    echo '  "SKIP_HOST_UPDATE": true' >> $HOME/.config/discord/settings.json
+    echo '}' >> $HOME/.config/discord/settings.json
+else
+    if [[ -z $(cat $HOME/.config/discord/settings.json | grep SKIP_HOST_UPDATE) ]]; then
+        sed -i '2i \  "SKIP_HOST_UPDATE": true,' $HOME/.config/discord/settings.json
+    fi
+fi
+
+if test "$BASH"; then
+  if [[ $(which starship) == "/usr/bin/starship" ]]; then
+    eval "$(starship init bash)"
+    if [ ! -f $HOME/.config/starship.toml ]; then
+        if [ -f /usr/share/starship/starship.toml ]; then
+            cp /usr/share/starship/starship.toml $HOME/.config/
+        fi
+    fi
+  fi
+fi
