diff --git a/plugins/copr.py b/plugins/copr.py
index afb8324..aa3644f 100644
--- a/plugins/copr.py
+++ b/plugins/copr.py
@@ -462,7 +462,7 @@ Bugzilla. In case of problems, contact the owner of this repository.
             dist = linux_distribution()
         # Get distribution architecture
         distarch = self.base.conf.substitutions['basearch']
-        if "Fedora" in dist[0]:
+        if "Fedora" or "Nobara" in dist[0]:
             if "Rawhide" in dist:
                 chroot = ("fedora-rawhide-" + distarch)
             # workaround for enabling repos in Rawhide when VERSION in os-release
