diff --git a/meson.build b/meson.build
index 420b6f9..d336409 100644
--- a/meson.build
+++ b/meson.build
@@ -18,12 +18,7 @@
 # along with obs-vaapi. If not, see <http://www.gnu.org/licenses/>.
 #
 
-project('obs-vaapi', 'c',
-  default_options : [
-    'prefix=~/.config/obs-studio/plugins/obs-vaapi/bin/64bit',
-    'libdir='
-  ],
-)
+project('obs-vaapi', 'c')
 
 library('obs-vaapi',
 	'obs-vaapi.c',
@@ -42,4 +37,5 @@ library('obs-vaapi',
 	gnu_symbol_visibility : 'hidden',
 	name_prefix : '',
 	install : true,
+	install_dir : join_paths(get_option('libdir'), 'obs-plugins'),
 )
