From 59560f77c7c4446b74afdc944fa3a3fe9c4b277c Mon Sep 17 00:00:00 2001
From: GloriousEggroll <gloriouseggroll@gmail.com>
Date: Thu, 2 Oct 2025 01:13:40 -0600
Subject: [PATCH 3/3] change log threshold default to WARNING instead of INFO

---
 inc/enum.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/inc/enum.h b/inc/enum.h
index 69d08ef..f622f8e 100644
--- a/inc/enum.h
+++ b/inc/enum.h
@@ -177,7 +177,7 @@ enum LogThreshold {
 	WARNING,
 	ERROR,
 	FATAL,
-	LOG_THRESHOLD_DEFAULT = INFO,
+	LOG_THRESHOLD_DEFAULT = WARNING,
 };
 enum LogThreshold log_threshold_val(const char *name);
 const char *log_threshold_name(enum LogThreshold log_threshold);
-- 
2.55.0

