--- a/goverlay.lpi	2020-11-03 03:43:38.000000000 +0200
+++ b/goverlay.lpi	2020-11-03 10:51:07.449457330 +0200
@@ -35,7 +35,8 @@
           </CodeGeneration>
           <Linking>
             <Debugging>
-              <GenerateDebugInfo Value="False"/>
+              <GenerateDebugInfo Value="True"/>
+              <DebugInfoType Value="dsDwarf2"/>
               <StripSymbols Value="True"/>
             </Debugging>
             <LinkSmart Value="True"/>
