From ee6e42f37ef6cfb6bacf82cea75a5e2267a6455b Mon Sep 17 00:00:00 2001
From: Patrick Laimbock <patrick@laimbock.com>
Date: Mon, 3 Mar 2025 20:32:22 +0100
Subject: [PATCH 1/2] Change version to 5.1.2

---
 meson.build | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/meson.build b/meson.build
index 9e69128d..5e18720f 100644
--- a/meson.build
+++ b/meson.build
@@ -1,7 +1,7 @@
 project(
   'yabridge',
   'cpp',
-  version : '5.1.1',
+  version : '5.1.2',
   meson_version : '>=0.56',
   default_options : [
     'warning_level=3',
-- 
2.48.1

