diff --git a/sddm-theme/Login.qml b/sddm-theme/Login.qml
index 33cde06..397afbf 100644
--- a/sddm-theme/Login.qml
+++ b/sddm-theme/Login.qml
@@ -103,9 +103,6 @@ SessionManagementScreen {
             placeholderText: i18ndc("plasma-desktop-sddm-theme",  "@info:placeholder in textfield", "Password")
             focus: !showUsernamePrompt || lastUserName
 
-            // Disable reveal password action because SDDM does not have the breeze icon set loaded
-            rightActions: []
-
             onAccepted: {
                 if (root.loginScreenUiVisible) {
                     startLogin();
