From d393795af80cdecef13335c87d3bc3e1e4b51ed5 Mon Sep 17 00:00:00 2001
From: GloriousEggroll <gloriouseggroll@gmail.com>
Date: Sat, 2 Mar 2024 18:34:17 -0700
Subject: [PATCH] fix no longer needed password view disable

---
/sddm-theme/Login.qml | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/sddm-theme/Login.qml b/sddm-theme/Login.qml
index 4a09aea..2d94676 100644
--- a/sddm-theme/Login.qml
+++ b/sddm-theme/Login.qml
@@ -102,9 +102,6 @@ SessionManagementScreen {
             placeholderText: i18nd("plasma-desktop-sddm-theme", "Password")
             focus: !showUsernamePrompt || lastUserName
 
-            // Disable reveal password action because SDDM does not have the breeze icon set loaded
-            rightActions: []
-
             onAccepted: {
                 if (root.loginScreenUiVisible) {
                     startLogin();
-- 
2.44.0

