diff --git a/kcms/landingpage/landingpage_kdeglobalssettings.kcfg b/kcms/landingpage/landingpage_kdeglobalssettings.kcfg
index 6bcce1e..788d017 100644
--- a/kcms/landingpage/landingpage_kdeglobalssettings.kcfg
+++ b/kcms/landingpage/landingpage_kdeglobalssettings.kcfg
@@ -7,7 +7,7 @@
   <group name="General">
     <entry name="colorScheme" key="ColorScheme" type="String">
       <label>Color scheme name</label>
-      <default>BreezeLight</default>
+      <default>BreezeDark</default>
     </entry>
   </group>
   <group name="KDE">
