diff --git a/libobs/obs.c b/libobs/obs.c
index 4a8856e..5e3f45a 100644
--- a/libobs/obs.c
+++ b/libobs/obs.c
@@ -1231,6 +1231,11 @@ extern void log_system_info(void);
 static bool obs_init(const char *locale, const char *module_config_path,
 		     profiler_name_store_t *store)
 {
+	/* This is a hack to force OBS to think it's running on gnome
+	 * because KDE6 currently has a bug that adds random
+	 * ampersands to title bar text */
+	setenv("XDG_CURRENT_DESKTOP", "gnome", false);
+
 	obs = bzalloc(sizeof(struct obs_core));
 
 	pthread_mutex_init_value(&obs->audio.monitoring_mutex);
