From a3d95ec7f0beca0befa978c873103657e26b27ec Mon Sep 17 00:00:00 2001
From: Daniel Werner <dan.werner@gmail.com>
Date: Sun, 16 Apr 2023 14:23:37 -0700
Subject: [PATCH 1/2] choose channel with highest power

---
 transport/mt76.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/transport/mt76.c b/transport/mt76.c
index 1d0a3b5..f9a628d 100644
--- a/transport/mt76.c
+++ b/transport/mt76.c
@@ -629,7 +629,9 @@ static int xone_mt76_get_channel_power(struct xone_mt76 *mt,
 static int xone_mt76_evaluate_channels(struct xone_mt76 *mt)
 {
 	struct xone_mt76_channel *chan;
-	int i, err;
+	int i, err, pow = 0;
+
+	mt->channel = NULL;
 
 	memcpy(mt->channels, xone_mt76_channels, sizeof(xone_mt76_channels));
 
@@ -644,13 +646,20 @@ static int xone_mt76_evaluate_channels(struct xone_mt76 *mt)
 		err = xone_mt76_switch_channel(mt, chan);
 		if (err)
 			return err;
+		
+		/* pick the highest power channel seen first */
+		if (chan->power > pow) {
+			mt->channel = chan;
+			pow = chan->power;
+		}
 
 		dev_dbg(mt->dev, "%s: channel=%u, power=%u\n", __func__,
 			chan->index, chan->power);
 	}
 
 	/* the last channel might not be the best one */
-	mt->channel = chan;
+	if (mt->channel == NULL)
+		mt->channel = chan;
 
 	return 0;
 }

From e18d91106a6d921a2883f4325b01733ac77bb344 Mon Sep 17 00:00:00 2001
From: Daniel Werner <dan.werner@gmail.com>
Date: Sun, 16 Apr 2023 14:24:27 -0700
Subject: [PATCH 2/2] move comment

---
 transport/mt76.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/transport/mt76.c b/transport/mt76.c
index f9a628d..44d5cae 100644
--- a/transport/mt76.c
+++ b/transport/mt76.c
@@ -648,6 +648,7 @@ static int xone_mt76_evaluate_channels(struct xone_mt76 *mt)
 			return err;
 		
 		/* pick the highest power channel seen first */
+		/* the last channel might not be the best one */
 		if (chan->power > pow) {
 			mt->channel = chan;
 			pow = chan->power;
@@ -657,7 +658,6 @@ static int xone_mt76_evaluate_channels(struct xone_mt76 *mt)
 			chan->index, chan->power);
 	}
 
-	/* the last channel might not be the best one */
 	if (mt->channel == NULL)
 		mt->channel = chan;
 
