From f7dd1c738b6b47de6f4e5b74886ef13810c43306 Mon Sep 17 00:00:00 2001
From: Caleb Davis <caleb.davis@gmail.com>
Date: Thu, 5 Mar 2026 08:26:57 -0500
Subject: [PATCH 857/857] hwmon: (nct6775) Add "ROG STRIX X870-I GAMING WIFI"

Signed-off-by: Caleb Davis <caleb.davis@gmail.com>
---
 drivers/hwmon/nct6775-platform.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/hwmon/nct6775-platform.c b/drivers/hwmon/nct6775-platform.c
index c3a719aef1ac..e912b662dff8 100644
--- a/drivers/hwmon/nct6775-platform.c
+++ b/drivers/hwmon/nct6775-platform.c
@@ -1403,6 +1403,7 @@ static const char * const asus_msi_boards[] = {
 	"ROG STRIX X670E-E GAMING WIFI",
 	"ROG STRIX X670E-F GAMING WIFI",
 	"ROG STRIX X670E-I GAMING WIFI",
+	"ROG STRIX X870-I GAMING WIFI",
 	"ROG STRIX X870E-H GAMING WIFI7",
 	"ROG STRIX Z590-A GAMING WIFI",
 	"ROG STRIX Z590-A GAMING WIFI II",
-- 
2.53.0

