diff --git a/usr/share/kde-settings/kde-profile/default/xdg/kdeglobals b/usr/share/kde-settings/kde-profile/default/xdg/kdeglobals
index 80c1393..308709c 100644
--- a/usr/share/kde-settings/kde-profile/default/xdg/kdeglobals
+++ b/usr/share/kde-settings/kde-profile/default/xdg/kdeglobals
@@ -9,7 +9,7 @@ toolBarFont=Noto Sans,9,-1,5,50,0,0,0,0,0
 Theme=breeze
 
 [KDE]
-LookAndFeelPackage=org.fedoraproject.fedora.desktop
+LookAndFeelPackage=org.kde.breezedark.desktop
 SingleClick=false
-ColorScheme=BreezeLight
+ColorScheme=BreezeDark
 widgetStyle=Breeze
