diff --git a/CMakeLists.txt b/CMakeLists.txt
index db294d3..97883b6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,5 +40,4 @@ include_directories(SYSTEM
 
 set_target_properties(linux-pipewire-audio PROPERTIES PREFIX "")
 
-install(TARGETS linux-pipewire-audio LIBRARY DESTINATION ${CMAKE_INSTALL_PREFIX}/${CMAKE_INSTALL_LIBDIR}/obs-plugins)
-install(DIRECTORY data/locale DESTINATION ${CMAKE_INSTALL_PREFIX}/share/obs/obs-plugins/linux-pipewire-audio)
\ No newline at end of file
+install(TARGETS linux-pipewire-audio LIBRARY DESTINATION ${CMAKE_INSTALL_PREFIX}/lib/obs-plugins)
