diff --git a/gtk/gtksettings.c b/gtk/gtksettings.c
index fb303bb2e4..0477f778ae 100644
--- a/gtk/gtksettings.c
+++ b/gtk/gtksettings.c
@@ -588,7 +588,7 @@ gtk_settings_class_init (GtkSettingsClass *class)
    * Since: 4.6
    */
   pspecs[PROP_HINT_FONT_METRICS] = g_param_spec_boolean ("gtk-hint-font-metrics", NULL, NULL,
-                                                         FALSE,
+                                                         TRUE,
                                                          GTK_PARAM_READWRITE);
 
   /**
